File: typescript/hamming/hamming.ts
export default class Hamming {
  constructor() {};

  public compute(DNAStrand1: string, DNAStrand2: string): number {
    if(DNAStrand1.length != DNAStrand2.length) {
      throw new Error('DNA strands must be of equal length.');
    }

    let strand1Array: Array<string> = [...DNAStrand1];
    let strand2Array: Array<string> = [...DNAStrand2];

    return strand1Array.filter((nucleotide: string, index: number) => nucleotide !== strand2Array[index]).length;
  }
}

File: lua/hamming/hamming.lua
local Hamming = {}

function Hamming.compute(a,b)
  if #a ~= #b then
    return -1
  else
    local distance = 0
    for i = 1, #a do
      if string.sub(a, i, i) ~= string.sub(b, i, i) then
        distance = distance + 1
      end
    end
    return distance
  end
end

return Hamming
